Simplify ClientState

raboof opened the pull request:
As rediscovering/retrying is now the responsibility of the netty client

/cc @tayvs since you worked on retry logic, that should be intact and not rely on the retry logic in CiientState AFAICS

tayvs commented:
@raboof So you remove channel recreation logic? Now recreation would work on subchannel level and if 'super channel' failed, then `monitorChannel` would catch failure and close client, Am I right?

raboof commented:
> So you remove channel recreation logic?

I removed the recreation logic in ClientState, because in #811 we switched things around so that discovery is also done 'inside' the Channel. This means when discovery fails it no longer prevents creation of the Channel, so it doesn't need to be retried.

> Now recreation would work on subchannel level and if 'super channel' failed, then monitorChannel would catch failure and close client, Am I right?

I think the 'super channel' will now never fail anymore when there is a recoverable failure - do you know of a scenario where it would?

tayvs commented:
@raboof I investigated about cases where 'super channel' would fail and it seems that it would not fail. If subchannel fail then it would be rediscovered and recreated.

raboof commented:
Great, thanks for validating!

ennru commented on interop-tests/src/test/scala/akka/grpc/scaladsl/LoadBalancingIntegrationSpec.scala:
```diff
@@ -133,6 +133,28 @@ class LoadBalancingIntegrationSpec extends AnyWordSpec with Matchers with Before
       service2.greetings.get should be(requestsPerServer)
     }
+
+    "select the right endpoint among invalid ones" in {
```
Maybe add a test case without a valid address to validate failing discovery?

ennru commented on runtime/src/main/scala/akka/grpc/internal/ClientState.scala:
```diff
@@ -9,6 +9,7 @@ import java.util.concurrent.atomic.AtomicReference
+import akka.annotation.InternalStableApi
 import akka.event.LoggingAdapter
 import akka.grpc.GrpcClientSettings
 import akka.pattern.Patterns
```
Some imports can go now.

ennru commented on runtime/src/main/scala/akka/grpc/internal/ClientState.scala:
```diff
@@ -22,138 +23,35 @@ import scala.util.Failure
   }
 
-  // used from generated client
+  // TODO #761 consider not leaking ManagedChannel to the generated code?
```
This seems important. Is there a reason why users would need the "Managed" channel?

raboof replied:
I guess it was convenient at the time, but let's see if we can refactor that (in a different PR).
